### PR TITLE
flash: fix `flash_read` API for NXP LPC55S36

### LINKED ISSUE #80325: nxp/lpc55s36: failure reading from internal Flash via `flash_read()`
**Describe the bug**

I am having trouble reading from internal Flash on the LPC55S36 on current `main`. 
I traced the issue from `flash_read()` to `flash_mcux_read()` [calling](https://github.com/butok/zephyr/blob/main/drivers/flash/soc_flash_mcux.c#L210) the NXP HAL function [`FLASH_IsFlashAreaReadable`](https://github.com/zephyrproject-rtos/hal_nxp/blob/master/mcux/mcux-sdk/devices/LPC55S36/drivers/flash/src/fsl_flash.c#L290-L296), which jumps to an hardcoded ROM/bootloader code (not accessible by the debug probe) and in a few instructions to the given `addr`, causing a panic. 

**To Reproduce**

This is reproducible on my end building and running the unmodified `samples/drivers/flash_shell` sample on an LPCXpresso55S36 eval board:
```
west build -p -b lpcxpresso55s36 samples/drivers/flash_shell/
west flash -r linkserver
```

Executing `flash read flash-controller@34000 0 1` on the uart console that pops up triggers a fault and crashes Zephyr.

**Expected behavior**
A proper Flash read on this SOC. 

Note that a quick test of replacing [the body of `flash_mcux_read`](https://github.com/butok/zephyr/blob/main/drivers/flash/soc_flash_mcux.c#L193-238) with a simple `FLASH_Read()` HAL API call results in the expected behavior, including detection of non-available pages.

**Impact**
Blocker for Flash API access on this platform. 

**Logs and console output**
Running a valid command on the `flash_shell` sample shows the following console output:
```
Flash shell sample
*** Booting Zephyr OS build v3.7.0-4863-g185432c96dd7 ***
uart:~$ flash read flash-controller@34000 0 1
[00:00:09.488,000] <err> os: ***** USAGE FAULT *****
[00:00:09.488,000] <err> os:   Illegal use of the EPSR
[00:00:09.488,000] <err> os: r0/a1:  0x00000076  r1/a2:  0x40000000  r2/a3:  0x00000001
[00:00:09.488,000] <err> os: r3/a4:  0x13028c41 r12/ip:  0x0000bf4d r14/lr:  0x0000bf69
[00:00:09.488,000] <err> os:  xpsr:  0x80000000
[00:00:09.488,000] <err> os: Faulting instruction address (r15/pc): 0x00000000
[00:00:09.488,000] <err> os: >>> ZEPHYR FATAL ERROR 35: Unknown error on CPU 0
[00:00:09.488,000] <err> os: Current thread: 0x20000630 (shell_uart)
[00:00:09.551,000] <err> os: Halting system
```

Note the `Faulting instruction address (r15/pc): 0x00000000` changes to mimic the requested address (`0` in the command above). E.g. `flash read flash-controller@34000 0x1234 4` results in  `Faulting instruction address (r15/pc): 0x00001234`

The backtrace as I could capture is:

```
#0  z_arm_usage_fault () at zephyr/arch/arm/core/cortex_m/fault_s.S:80
#1  <signal handler called>
#2  thread_valid_cb (cthread=0x76 <_irq_vector_table+54>, user_data=0x40000000) at zephyr/subsys/shell/modules/kernel_service/thread/thread.c:31
#3  0x0000bf68 in flash_mcux_read (dev=<optimized out>, offset=<optimized out>, data=0x20009d2c <shell_uart_stack+1628>, len=23567)
    at zephyr/drivers/flash/soc_flash_mcux.c:210
#4  0x0000d7d4 in shell_uart_backend ()
```

(`thread_valid_cb` is the symbol closest to address 0 AFAICS). 

**Environment (please complete the following information):**

 - OS: Linux
 - Toolchain: Zephyr SDK 0.16.8
 - Commit SHA: `185432c96dd7`

### PR DESCRIPTION
**Pull request by pillo79**
This PR reworks the LPC55S36 implementation of the `flash_mcux_read()` function to work around issue #80325. Feedback welcome!

This is achieved on the `CONFIG_CHECK_BEFORE_READING` path (the default) by using the `FLASH_Read()` HAL API call, which is resilient to errors, and map the error codes to properly emulate an erased page read. 

When `CONFIG_CHECK_BEFORE_READING` is disabled, the code reverts to the previous `memcpy()` operation, which is faster but unsafe on erased or otherwise unreadable pages.

`#ifdef` guards kept in the style of the file, though some may be changed to `if (IS_ENABLED(...))` if desired.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/80325

### CODE REVIEW

**Comment by pillo79** [comment]
Friendly ping for any feedback after a week of silence :slightly_smiling_face:  
I know Zephyr is in freeze right now, but since this is a bugfix I think this would be useful in 4.0.

File: drivers/flash/soc_flash_mcux.c [hunk 1]
```diff
@@ -204,36 +204,49 @@ static int flash_mcux_read(const struct device *dev, off_t offset,
...
-#ifdef CONFIG_CHECK_BEFORE_READING
-  #ifdef CONFIG_SOC_LPC55S36
-	/* Validates the given address range is loaded in the flash hiding region. */
-	rc = FLASH_IsFlashAreaReadable(&priv->config, addr, len);
-	if (rc != kStatus_FLASH_Success) {
-		rc = -EIO;
-	} else {
-		/* Check whether the flash is erased ("len" and "addr" must be word-aligned). */
-		rc = FLASH_VerifyErase(&priv->config, ((addr + 0x3) & ~0x3),  ((len + 0x3) & ~0x3));
+#if defined(CONFIG_CHECK_BEFORE_READING) && !defined(CONFIG_SOC_LPC55S36)
```
Comment (danieldegrasse) [inline drivers/flash/soc_flash_mcux.c:207]: A suggestion here for readability- could we do something like this:

```
if (IS_ENABLED(CONFIG_CHECK_BEFORE_READING)) {
    if (IS_ENABLED(CONFIG_SOC_LPC55S36) {
        /* LPC55S36 specific logic */
    } else {
       rc = is_area_readable(addr, len)
    }
}

if (!rc) {
   memcpy(data, (void *)addr, len);
} else if ((rc == -ENODATA) && IS_ENABLED(CONFIG_CHECK_BEFORE_READING)) {
  /* Erased area, return dummy data as an erased page. */
  memset(data, 0xFF, len);
  rc = 0;
}
return rc;
```

Apologies if formatting there is confusing. At least IMO, the current function implementation in `main` is hard to follow, and I figure if we are fixing the bug cleaning up the readability wouldn't hurt either

Comment (pillo79) [reply drivers/flash/soc_flash_mcux.c:207]: Completely agree, but it's really a thorny situation. The `FLASH_Read` does a "copy before verify" which doesn't neatly fit in the previous model:

| `CHECK` is set? | on `55S36`? | Check stage | Copy stage | `0xff` fill 
|--------|--------|--------|-------|-------|
| no | - | no | `memcpy` | no |
| yes| no | `is_area_readable` | `memcpy` | yes |
| yes| yes| `FLASH_Read` + `VerifyErased` | (already done) | yes

My first proposal avoided any duplicate code, but ended up creating awkward logical groupings.

**Comment by pillo79** [comment]
**v2**:
* converted `#ifdef ...` to `if (IS_ENABLED(...))` 
* spelled out all cases to make it way easier to read

File: drivers/flash/soc_flash_mcux.c (hunk 1 above)
Comment (danieldegrasse) [reply drivers/flash/soc_flash_mcux.c]: Thank you for the explanation here, this makes more sense (and I really like the rework in the latest push, it is a lot easier to follow)

File: drivers/flash/soc_flash_mcux.c [hunk 2]
```diff
@@ -204,36 +207,54 @@ static int flash_mcux_read(const struct device *dev, off_t offset,
...
-		/* Check whether the flash is erased ("len" and "addr" must be word-aligned). */
-		rc = FLASH_VerifyErase(&priv->config, ((addr + 0x3) & ~0x3),  ((len + 0x3) & ~0x3));
-		if (rc == kStatus_FLASH_Success) {
-			rc = -ENODATA;
+	if (IS_ENABLED(CONFIG_CHECK_BEFORE_READING)) {
+		/*
+		 * Ensure the area is readable, since a direct access may cause faults on erased
+		 * or otherwise unreadable pages. Emulate erased pages, return other errors.
+		 */
+		if (IS_ENABLED(CONFIG_SOC_LPC55S36)) {
```
Comment (danieldegrasse) [inline drivers/flash/soc_flash_mcux.c:215]: @pillo79 it looks like we may need more "#ifdefery", because some older NXP platforms don't define the `FLASH_Read` API at all. I think we probably need to use `#ifdef` here. Sorry for the poor guidance

**Comment by danieldegrasse** [comment]
On the next push can you reference the github issue within the commit message? IE add

`Fixes #80325` to the commit message (at the end, before the signed-off-by-line)

**Comment by pillo79** [comment]
Minor addition I spotted only now - fixed the alignment logic for `FLASH_VerifyErase`, which was too simple and did not properly enlarge the area.

**Review by danieldegrasse — APPROVED** [review]

**Comment by danieldegrasse** [comment]
@de-nordic can you take a look?

File: drivers/flash/soc_flash_mcux.c [hunk 3]
```diff
@@ -205,35 +205,55 @@ static int flash_mcux_read(const struct device *dev, off_t offset,
...
+	rc = FLASH_Read(&priv->config, addr, data, len);
+	switch (rc) {
+	case kStatus_FLASH_Success:
+		rc = 0;
+		break;
+	case kStatus_FLASH_EccError:
+		/* Check id the ECC issue is due to the Flash being erased
+		 * ("addr" and "len" must be word-aligned for this call).
+		 */
+		rc = FLASH_VerifyErase(&priv->config,
```
Comment (de-nordic) [inline drivers/flash/soc_flash_mcux.c:224]: Question: wouldn't it be better to just ifdef the `is_area_readable` internally to use this method of verification, whether memory is readable, for `CONFIG_SOC_LPC55S36`, and then follow it with the memcopy or memset depending on result? Instead of reading first and then checking why it failed, and have streamlined code fall all devices supported?

As far as I understand the description for `FLASH_VerifyErase` in fsl_iap.h, it is able to tell whether there is ECC problem in range (btw NXP, the function no longer takes margin as parameter, yet you still mention it in description).

Additional note: IMHO making read check configurable makes no sense, because there is always a chance that ECC error will topple device instead of returning with error, unless of course you disable ECC,

**Review by de-nordic — APPROVED** [review]
Looks OK, as far as I can tell.
Left a comment on potential place for streamlining execution here https://github.com/zephyrproject-rtos/zephyr/pull/80351/files#r1831641313, but that is up to the vendor here.
